'''
This is the main script that has to be executed.
Includes all the Snakefiles needed to execute the workflow.
'''

#Configuration yaml.
configfile: "config/config.yaml"


#Global variable
data_dir = config["data_dir"]
results_dir = config["results_dir"]
barcodes = config["barcodes.txt"]
ref_genome = config["ref_genome"]


#Global functions

def get_sample_names():
    sample_names = ["unknown_sample"]
    with open(f"{data_dir}/barcodes.txt", "r") as barcode_file:
        for line in barcode_file:
            split_line = line.strip().split(" ")
            sample_names.append(split_line[1].split("_R")[0])
    return sample_names


#Included rules
include: "Rules/fasta.smk"
include: "Rules/Mapping.smk"
include: "Rules/Trimming.smk"
include: "Rules/Variant-calling.smk"


#Output is the variant file and the plotted variants.
rule all:
    input:
        "results/variants.vcf",
        "results/variant_plots.png"
