'''
This is the main script that has to be executed.
Includes all the Snakefiles needed to execute the workflow.
'''

#Configuration yaml.
configfile: "config/config.yaml"


#Global variables
data_dir = config["data_dir"]
results_dir = config["results_dir"]
# barcodes = config["barcodes.txt"]
ref_genome = config["ref_genome"]
ref_genome_ext = config["ref_genome_ext"]
sample_name = config["sample_name"]
sample_ext = config["sample_ext"]


#Global functions

# def get_sample_names():
#     sample_names = ["unknown_sample"]
#     with open(f"{data_dir}/barcodes.txt", "r") as barcode_file:
#         for line in barcode_file:
#             split_line = line.strip().split(" ")
#             sample_names.append(split_line[1].split("_R")[0])
#     return sample_names


#Included rules
include: "Rules/Fasta.smk"
include: "Rules/Mapping.smk"
include: "Rules/Trimming.smk"
include: "Rules/Variant-calling.smk"
include: "Rules/Graph_plotting.smk"


#Output is the variant file and the plotted variants.
rule all:
    input:
        f"{results_dir}/pileups/{sample_name}.pileup.vcf",
        f"{results_dir}/variants.png"
